Refactor Shipyard to be consumable by other projects

mkolesnik opened the pull request:
These commits refactor Shipyard and break it down into smaller more basic components which are Submariner agnostic.
This allows it to be used in projects outside the Submariner org, i.e. https://github.com/skupperproject/skupper/pull/1317
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

mkolesnik committed:
Genericize base dapper logic

Make base dapper logic project agnostic so that it can be consumed by
projects outside the submariner org.
This mainly involves moving out basic dapper functions out to
`Makefile.dapper.base` and adding an `ORG` argument to specify the
organization.
References to `submariner-io` have been removed where possible.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Split out linting to its own Makefile

To make it easy for other projects to consume, split out linting to a
`Makefile.linting` file that can be included by other projects that
desire these capabilities.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Split out clusters to its own Makefile

To make it easy for other projects to consume, split out cluster
creation and cleanup to a `Makefile.clusters` file that can be
included by other projects that desire these capabilities.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Rename `GLOBALNET` to `OVERLAPPING`

To better support external projects, rename the flag to a more neutral
one which actually reflects the type of deployment.

The old flag (when sent via `using=`) is still respected, to not break
existing Submariner projects.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Make Shipyard's location dynamic

For easier consumption from other projects, and for easier POCs, make
the URL and image locations dynamic so that we can specify where to take base
Shipyard files and images from.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Move kindnet workaround to deploy script

Since the workaround is Submariner specific, it should be part of the
deploy stage and not part of the cluster creation stage which is
agnostic to Submariner.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Allow running only Shipyard goals in container

Add the option to run just Shipyard's goals inside the container, to
not interfere with consuming project's goals.

Currently, this only covers the basic clusters and Go linting goals, but
more goals can be added in the future.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

skitt commented on scripts/shared/deploy.sh:
```diff
@@ -123,6 +142,7 @@ declare_kubeconfig
 
 load_library deploy DEPLOYTOOL
 deploytool_prereqs
+[[ "$PROVIDER" == kind ]] || run_all_clusters schedule_dummy_pod
```
Isn’t this the wrong way round? If `schedule_dummy_pod` is supposed to run with kind,

```
[[ "$PROVIDER" != kind ]] || run_all_clusters schedule_dummy_pod
```

mkolesnik committed:
Add ability to specify custom config

Add an ability to specify custom config in the cluster config file,
which will be provider specific, and allow consuming projects to
customize the provider's configuration before deploying.

Currently, only 'kind' will be supported.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

skitt committed:
Merge branch 'devel' into remixed

skitt committed:
Merge branch 'devel' into remixed

tpantelis committed:
Merge branch 'devel' into remixed
